### PR TITLE
Document k6/experimental/redis module

### PR DESCRIPTION
**Pull request by oleiade**
This Pull Request adds a `/javascript-api` documentation section detailing the usage of the new `k6/experimental/redis` module to land in `v0.40`.

Entry points for the PR are the `src/data/markdown/docs/02\ javascript\ api/13\ k6-redis.md` and `src/data/markdown/docs/02\ javascript\ api/13\ k6-redis/10 Client.md` files.

PS: I'm aware the eslint doesn't pass, for a reason that is beyond my mere human understanding. Playing some didgeridoo to it, and bringing my strongest incantations to the table to try and get it to magically resolve 🤞🏻

### CODE REVIEW

File: src/data/markdown/docs/02 javascript api/13 k6-redis.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis.md:14]:
```suggestion
The [Redis](https://redis.io/) module for [k6](https://github.com/grafana/k6) provides a client library to interact with Redis, directly from their k6 scripts.
Its objective is to let users to either load test their Redis infrastructure, or use Redis as a data store for their test-script logic.
```

Comment (MattDodsonEnglish) [reply src/data/markdown/docs/02 javascript api/13 k6-redis.md:14]: I don't have a strong reason to remove "load". I did it for a few reasons:
- k6 could be for general performance testing, too
- it shortens the sentence
- I think style guidelines would say it needed to be "load-test-script logic", and while this would be very pleasing to me, I think it is too fussy for our docs.

File: src/data/markdown/docs/02 javascript api/13 k6-redis.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis.md:16]:
```suggestion
Though the API intends to be quite thorough and extensive, it is still limited to a set of functionalities that the k6 team judged useful in the context of k6.
Instead of a full-fledged Redis client, integrating the whole Redis API, the modules aims to empower users to use Redis for use cases most appropriate to k6. 
```

File: src/data/markdown/docs/02 javascript api/13 k6-redis.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis.md:23]:
```suggestion
The `Client` exposes a [promise](https://javascript.info/promise-basics)-based API.
Unlike most other current k6 modules and extensions,
which operate in a synchronous manner,
```

File: src/data/markdown/docs/02 javascript api/13 k6-redis.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis.md:24]:
```suggestion
the Redis `Client` operates in an asynchronous manner.
In practice, this means that using the Redis `Client`'s methods won't block test execution,
and that the test will continue to run even if the Redis `Client` isn't ready to respond to the request.
```

File: src/data/markdown/docs/02 javascript api/13 k6-redis.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis.md:26]:
```suggestion
This new execution model does introduce a potential caveat: to depend on operations against Redis, all following operations should be made in the context of a [promise chain](https://javascript.info/promise-chaining).
As the examples below demonstrate, whenever there is a dependency on a Redis operation result or return value, the operation should be wrapped in a promise itself.
This way, a user can perform asynchronous interactions with Redis in a seemingly synchronous manner.
```

File: src/data/markdown/docs/02 javascript api/13 k6-redis.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis.md:32]: > Client exposing allowing interactions with Redis.

Not sure here. Maybe:

Client that exposes allowed interactions with Redis.

File: src/data/markdown/docs/02 javascript api/13 k6-redis/10 Client.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis/10 Client.md:6]:
```suggestion
`Client` is a Redis client to interact with a Redis server or cluster. It exposes a promise-based API, which users can interact with in an asynchronous manner.

Though the `Client` API intends to be thorough and extensive, it does not expose the whole Redis API.
Instead, the API aims to empower users to use Redis for use cases most appropriate to k6. 
```

File: src/data/markdown/docs/02 javascript api/13 k6-redis/10 Client.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis/10 Client.md:8]:
```suggestion
Note that you configure the `Client` through the [`Options`](/javascript-api/k6-redis/options) object.
```

File: src/data/markdown/docs/02 javascript api/13 k6-redis.md
Comment (oleiade) [reply src/data/markdown/docs/02 javascript api/13 k6-redis.md:19]: The first sentence in your suggestion should drop the "their" from "directly from their k6 scripts". As the users it referred to is now gone 😄

File: src/data/markdown/docs/02 javascript api/13 k6-redis.md
Comment (oleiade) [reply src/data/markdown/docs/02 javascript api/13 k6-redis.md:44]: You're right, direct form is probably better here indeed!

**Comment by oleiade** [comment]
@MattDodsonEnglish The underlying type of [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) in JavaScript is actually an `Object`.
As a convenience, and courtesy to the developers, the `Promise<T>` signature are [Typescript](https://www.typescriptlang.org/) type annotations. It's just much easier to know what to expect when presented with the `Promise<number>` type, rather than `Object` and a long explanation that the object is actually a promise and yaddi yadda 😄 

 I'm working on a PR to [Definitely Typed](https://github.com/DefinitelyTyped/DefinitelyTyped) to also get type annotations for the Redis module in code editors via the `@types/k6` package 😇

**Review by javaducky — CHANGES_REQUESTED** [review]
Looks like we're missing the `k6/experimental/redis` summary and righ-links in the main k6 API page:

image

Guessing we won't want to merge this into the documentation until v0.40 is cut, unless we're able to hide the links from the left-hand gutter until that point using `hideFromSidebar` attribute.

Another potential thing to consider is that we may want to create the v0.39 versioned api directory before this is merged.

**Comment by MattDodsonEnglish** [comment]
Thanks to Paul for looking at the formatting, I haven't even looked at that.

Here's a link to the WiP in staging, by the way: https://mdr-ci.staging.k6.io/docs/refs/pull/768/merge/javascript-api/k6-experimental-redis/

File: src/data/markdown/docs/02 javascript api/13 k6-redis.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis.md:20]:
```suggestion
The [Redis](https://redis.io/) module provides a client library that makes it possible to interact with Redis directly from a k6 script.
With this module, you can load test Redis or use Redis as a data store for test-script logic.
```

File: src/data/markdown/docs/02 javascript api/13 k6-redis.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis.md:25]:
```suggestion
This module is **not stable**, and remains open to changes. So, if you have a use case for some Redis functionality that's missing, feel free to open an issue or pull request.
```

File: src/data/markdown/docs/02 javascript api/13 k6-redis/20 Options.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-redis/20 Options.md:6]:
```suggestion
You can configure the [Redis Client](/javascript-api/k6-redis/client) at construction time with the [Options](#options) object.
We recommend passing the options to the constructor as an argument,  then passing the most common options, such as the `addrs` and `password`, to the constructor from the environment.

The following snippet provides an example:
```

**Review by MattDodsonEnglish — COMMENTED** [review]
Very nice, and extensive work.

All my edits are to trim the language down, or remove directional language. You might notice that I'm breaking lines by sentence. This is just a minor improvment to make the raw text a bit easier to read, and diffs easier to look at in the future. Not worth changing everywhere.

To be honest, I did not read all of the many, many, options descriptions, but the ones I did all seemed descriptive, grammatical, and sufficiently spartan. :+1: 

I have some comments about the prefatory table about return values:

- Why **bold**  "success", "resolve" and "reject"? I think readers will most know that information, since it's obvious from the table headers. I'd vote for not bolding
- "Error on" seems a little off, Maybe "Error when" or "Error if" is more appropriate.
- You may have seen in some internal docs that I am starting to have an existential crisis about tables. They just don't work well on narrow screens. (Gov.uk bans tables for all but figures).

This got me thinking of more vertical ways to orient the content. Since the tables have only two rows, I think you could transpose the rows into columns. In the screenshot, I'm presenting three options:

image

I don't really have a strong preference---I've put in many many-columned tables too---but I wanted to show other ways.

The introduction still seems to have a little too much prose for me, but I'm not sure what to cut or break up. Something for the next iteration. Overall very nice set of reference docs, if you ask me. Great examples!

**Comment by MattDodsonEnglish** [comment]
One other structural thing I forgot to mention: I believe this doc is
generated from a data file, right? If it is, I'd prefer if there were a
flag that skips table rows when the value is empty. We have a few "on
error" columns that have no content.

Wouldn't it be better just to remove the column?


On Thursday, August 11, 2022, Paul Balogh ***@***.***> wrote:
> @javaducky requested changes on this pull request.
>
> Looks like we're missing the k6/experimental/redis summary and righ-links
in the main k6 API page:
>
> <
https://ci5.googleusercontent.com/proxy/mepACI0yMuGSkWN1X0os7vjVeHnyCf6DTfS-cVYm8T2eGNBzTDp9iBr_hWhPkdizUoa6FILSo3d89i3fQcXEqsAICb7itsmhcgJIEni-YlNjt93zzIk-1ipLeC_mOu6vq1KwMNKPHV7YXVYfmjFj3PH9e_6zbg=s0-d-e1-ft#https://user-images.githubusercontent.com/944496/184231191-f5fb614e-5500-44e6-8b4c-8aa6420ad979.png
>
>
> Guessing we won't want to merge this into the documentation until v0.40
is cut, unless we're able to hide the links from the left-hand gutter until
that point using hideFromSidebar attribute.
>
> Another potential thing to consider is that we may want to create the
v0.39 versioned api directory before this is merged.
>
> ________________________________
>
> In src/data/markdown/docs/02 javascript api/13 k6-redis/10 Client.md:
>
>> +## List methods
> +
> +| Method | Redis command | Description |
> +|
:-------------------------------------------------------------------------------
| :----------------------------------------------- |
:------------------------------------------------------------------------------
|
> +| [`Client.lpush(key,
values)`](/javascript-api/k6-redis/client/client-lpush) | **[LPSUH](
https://redis.io/commands/lpsuh)** | Inserts all the specified values at
the head of the list stored at `key`. |
> +| [`Client.rpush(key,
values)`](/javascript-api/k6-redis/client/client-rpush) | **[RPUSH](
https://redis.io/commands/rpush)** | Inserts all the specified values at
the tail of the list stored at `key`. |
> +| [`Client.lpop(key)`](/javascript-api/k6-redis/client/client-lpop) |
**[LPOP](https://redis.io/commands/lpop)** | Removes and returns the first
element of the list stored at `key`. |
> +| [`Client.rpop(key)`](/javascript-api/k6-redis/client/client-rpop) |
**[RPOP](https://redis.io/commands/rpop)** | Removes and returns the last
element of the list stored at `key`. |
> +| [`Client.lrange(key, start,
stop)`](/javascript-api/k6-redis/client/client-lrange) | **[LRANGE](
https://redis.io/commands/lrange)** | Returns the specified elements of the
list stored at `key`. |
> +| [`Client.lindex(key, start,
stop)`](/javascript-api/k6-redis/client/client-lindex) | **[LINDEX](
https://redis.io/commands/lindex)** | Returns the specified element of the
list stored at `key`. |
> +| [`Client.lset(key, index,
element)`](/javascript-api/k6-redis/client/client-lset) | **[LSET](
https://redis.io/commands/lset)** | Sets the list element at `index` to
`element`. |
> +| [`Client.lrem(key, count,
value)`](/javascript-api/k6-redis/client/client-lrem) | **[LREM](
https://redis.io/commands/lrem)** | Removes the first `count` occurrences
of `value` from the list stored at `key`. |
> +| [`Client.llen(key)`](/javascript-api/k6-redis/client/client-llen) |
**[LLEN](https://redis.io/commands/llen)** | Returns the length of the list
stored at `key`. |
> +
> +
> +### Hash methods
>
> A note on the consistency. This grouping (and most other groupings) uses
### whereas the "List methods" uses ##. If you put them all as 2-hashes,
Gatsby will create the links in the right-gutter. As it is now, only Lists
are included.
>
> <
https://ci5.googleusercontent.com/proxy/IaIRSOyVhr3ayRLLWQTUgeUcKyhNRQvJOvWTmtCMIVWTeLlwnI6urQRYYY_AvwGPIOpU6YKuMJPZW2GckDq0BQmkwy5-MyfrgmfxLlwS2rt47Pmz2-W0l4rb0AeVLIRk3sh9JpzLjFXhN3CfFUov_VidxvRs6g=s0-d-e1-ft#https://user-images.githubusercontent.com/944496/184221791-e7ba1690-d01f-4f0a-bc27-2e63bff5eb35.png
>
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.<
https://ci4.googleusercontent.com/proxy/47oLC7SOdJyakvxZ7RrFNltjS3_549M1aWull6l81yyA2wVFNwL4Ttct0sfCC2pw9gqLXW0rzdIRMeCeSvjtvdysai5liwvnDJdtu1Zs6btDwzWZXkS6GSqGGYRJ4QdIAKWhBobtT03SNFU1L341cuYbBRR01NnaFO0IAiSN_EXDr95a3OZ_NxQKYjgXtHlX9mflDlqiwJ5267qIMYpxHuGYMVRZSrIU6hoS1M3eJUryZHQ4tZnd=s0-d-e1-ft#https://github.com/notifications/beacon/ALJQUZGP4E7BAFAUJZITWMTVYVNCDA5CNFSM56HISB7KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOH7FJG5A.gif>Message
ID: ***@***.***>

**Comment by oleiade** [comment]
@MattDodsonEnglish

Regarding empty columns, I'd prefer to keep them, to explicit the fact that a promise **always** expose a `resolve` callback (that's called on success), and a `rejected` callback (which is called on failure/error). Promises are quite new in the k6 ecosystem, and both for educational, and consistency purposes, I think it's better to have a visual reminder that the promise might be rejected. Moreover, although sometimes there are no direct causes for rejection, the promise might still be rejected by external factors ("can't reach the Redis server").

Which leads into the way we structure the table. I believe we could name the promise state related columns differently **on resolution**, and **on rejection**, instead of the current naming. I want to insist on the "on" prefix there, to emphasize the asynchronousness of it. Also, the reason I had **success** and **rejection** in bold, was to have one more visual cue, but it might not be useful indeed.

**Comment by oleiade** [comment]
@javaducky 

Thanks for the review, much appreciated 🙇🏻 

We won't merge this before the `v0.40.0` release, indeed.

**Comment by oleiade** [comment]
@MattDodsonEnglish I've updated the tables to be as follows:
image
Let me know what you think :)

**Comment by oleiade** [comment]
@javaducky I've updated the whole structuring of the titling, and renamed a few files. The structure was indeed broken. It now should look conform to the rest, let me know if you spot any leftovers 🙇🏻

File: src/data/markdown/docs/02 javascript api/12 k6-experimental-redis.md
Comment (javaducky) [inline src/data/markdown/docs/02 javascript api/12 k6-experimental-redis.md:47]:
```suggestion
| [Client](/javascript-api/k6-experimental-redis/client)   | Client that exposes allowed interactions with Redis.                                           |
```
You have a broken link.

File: src/data/markdown/docs/02 javascript api/12 k6-experimental-redis.md
Comment (javaducky) [inline src/data/markdown/docs/02 javascript api/12 k6-experimental-redis.md:48]:
```suggestion
| [Options](/javascript-api/k6-experimental-redis/options) | Options used to configure the behavior of the [Redis Client](/javascript-api/k6-experimental-redis/client). |
```
Broken links.

File: src/data/markdown/docs/02 javascript api/13 k6-experimental-redis/10 Client.md
Comment (javaducky) [inline src/data/markdown/docs/02 javascript api/13 k6-experimental-redis/10 Client.md:11]:
```suggestion
Note that the `Client` is configured through the [`Options`](/javascript-api/k6-experimental-redis/options) object.
```
Broken link.

File: src/data/markdown/docs/02 javascript api/13 k6-experimental-redis/20 Options.md
Comment (javaducky) [inline src/data/markdown/docs/02 javascript api/13 k6-experimental-redis/20 Options.md:6]:
```suggestion
You can configure the [Redis Client](/javascript-api/k6-experimental-redis/client) at construction time with the [Options](#options) object.
```
Broken link.

**Review by javaducky — CHANGES_REQUESTED** [review]
There's a few broken links given the refactor of the landing page name.

**Comment by MattDodsonEnglish** [comment]
> @MattDodsonEnglish I've updated the tables to be as follows: image Let me know what you think :)

I like it a lot more :+1:. Much more precise. 

For alternative presentations of the data, I can only think of Vertical columns and lists. I'm pretty much indifferent about what's best.

I made this test, but the screenshot was so massive it was hard to see the differences. Hopefully this time the comparison is clearer:

Again, these are just alternatives; I don't have a strong opinion.
image

**Review by javaducky — APPROVED** [review]
Looks good to merge once we cut release v0.40.0.

**Comment by oleiade** [comment]
@MattDodsonEnglish Thanks for pointing out the alternatives. I think I will go with the vertical columns 👍🏻

File: src/data/markdown/docs/02 javascript api/13 k6-experimental-redis/10 Client.md
Comment (MattDodsonEnglish) [inline src/data/markdown/docs/02 javascript api/13 k6-experimental-redis/10 Client.md:182]:
```suggestion
| [`Client.set(key, value, expiration)`](/javascript-api/k6-experimental-redis/client/client-set) | **[SET](https://redis.io/commands/set)**             | Set `key` to hold `value`, with a time to live equal to `expiration`.       |
| [`Client.get(key)`](/javascript-api/k6-experimental-redis/client/client-get)                    | **[GET](https://redis.io/commands/get)**             | Get the value of `key`.                                                     |
| [`Client.getSet(key, value)`](/javascript-api/k6-experimental-redis/client/client-getset)       | **[GETSET](https://redis.io/commands/getset)**       | Atomically sets `key` to `value` and returns the old value stored at `key`. |
| [`Client.del(keys)`](/javascript-api/k6-experimental-redis/client/client-del)                   | **[DEL](https://redis.io/commands/del)**             | Removes the specified keys.                                                 |
| [`Client.getDel(key)`](/javascript-api/k6-experimental-redis/client/client-getdel)              | **[GETDEL](https://redis.io/commands/getdel)**       | Get the value of `key` and delete the key.                                  |
| [`Client.exists(keys)`](/javascript-api/k6-experimental-redis/client/client-exists)             | **[EXISTS](https://redis.io/commands/exists)**       | Returns the number of `key` arguments that exist.                           |
| [`Client.incr(key)`](/javascript-api/k6-experimental-redis/client/client-incr)                  | **[INCR](https://redis.io/commands/incr)**           | Increments the number stored at `key` by one.                               |
| [`Client.incrBy(key, increment)`](/javascript-api/k6-experimental-redis/client/client-incrby)   | **[INCRBY](https://redis.io/commands/incrby)**       | Increments the number stored at `key` by `increment`.                       |
| [`Client.decr(key)`](/javascript-api/k6-experimental-redis/client/client-decr)                  | **[DECR](https://redis.io/commands/decr)**           | Decrements the number stored at `key` by one.                               |
| [`Client.decrBy(key, decrement)`](/javascript-api/k6-experimental-redis/client/client-decrby)   | **[DECRBY](https://redis.io/commands/decrby)**       | Decrements the number stored at `key` by `decrement`.                       |
| [`Client.randomKey()`](/javascript-api/k6-experimental-redis/client/client-randomkey)           | **[RANDOMKEY](https://redis.io/commands/randomkey)** | Returns a random key's value.                                               |
| [`Client.mget(keys)`](/javascript-api/k6-experimental-redis/client/client-mget)                 | **[MGET](https://redis.io/commands/mget)**           | Returns the values of all specified keys.                                   |
| [`Client.expire(key, seconds)`](/javascript-api/k6-experimental-redis/client/client-expire)     | **[EXPIRE](https://redis.io/commands/expire)**       | Sets a timeout on key, after which the key will automatically be deleted.   |
| [`Client.ttl(key)`](/javascript-api/k6-experimental-redis/client/client-ttl)                    | **[TTL](https://redis.io/commands/ttl)**             | Returns the remaining time to live of a key that has a timeout.             |
| [`Client.persist(key)`](/javascript-api/k6-experimental-redis/client/client-persist)            | **[PERSIST](https://redis.io/commands/persist)**     | Removes the existing timeout on key.                                        |

## List methods

| Method                                                                           | Redis command                                    | Description                                                                     |
| :------------------------------------------------------------------------------- | :----------------------------------------------- | :------------------------------------------------------------------------------ |
| [`Client.lpush(key, values)`](/javascript-api/k6-experimental-redis/client/client-lpush)        | **[LPSUH](https://redis.io/commands/lpsuh)**   | Inserts all the specified values at the head of the list stored at `key`.       |
| [`Client.rpush(key, values)`](/javascript-api/k6-experimental-redis/client/client-rpush)        | **[RPUSH](https://redis.io/commands/rpush)**   | Inserts all the specified values at the tail of the list stored at `key`.       |
| [`Client.lpop(key)`](/javascript-api/k6-experimental-redis/client/client-lpop)                  | **[LPOP](https://redis.io/commands/lpop)**     | Removes and returns the first element of the list stored at `key`.              |
| [`Client.rpop(key)`](/javascript-api/k6-experimental-redis/client/client-rpop)                  | **[RPOP](https://redis.io/commands/rpop)**     | Removes and returns the last element of the list stored at `key`.               |
| [`Client.lrange(key, start, stop)`](/javascript-api/k6-experimental-redis/client/client-lrange) | **[LRANGE](https://redis.io/commands/lrange)** | Returns the specified elements of the list stored at `key`.                     |
| [`Client.lindex(key, start, stop)`](/javascript-api/k6-experimental-redis/client/client-lindex) | **[LINDEX](https://redis.io/commands/lindex)** | Returns the specified element of the list stored at `key`.                      |
| [`Client.lset(key, index, element)`](/javascript-api/k6-experimental-redis/client/client-lset)  | **[LSET](https://redis.io/commands/lset)**     | Sets the list element at `index` to `element`.                                  |
| [`Client.lrem(key, count, value)`](/javascript-api/k6-experimental-redis/client/client-lrem)    | **[LREM](https://redis.io/commands/lrem)**     | Removes the first `count` occurrences of `value` from the list stored at `key`. |
| [`Client.llen(key)`](/javascript-api/k6-experimental-redis/client/client-llen)                  | **[LLEN](https://redis.io/commands/llen)**     | Returns the length of the list stored at `key`.                                 |


## Hash methods

| Method                                                                                  | Redis command                                      | Description                                                                                          |
| :-------------------------------------------------------------------------------------- | :------------------------------------------------- | :--------------------------------------------------------------------------------------------------- |
| [`Client.hset(key, field, value)`](/javascript-api/k6-experimental-redis/client/client-hset)           | **[HSET](https://redis.io/commands/hset)**       | Sets the specified field in the hash stored at `key` to `value`.                                     |
| [`Client.hsetnx(key, field, value)`](/javascript-api/k6-experimental-redis/client/client-hsetnx)       | **[HSETNX](https://redis.io/commands/hsetnx)**   | Sets the specified field in the hash stored at `key` to `value`, only if `field` does not yet exist. |
| [`Client.hget(key, field)`](/javascript-api/k6-experimental-redis/client/client-hget)                  | **[HGET](https://redis.io/commands/hget)**       | Returns the value associated with `field` in the hash stored at `key`.                               |
| [`Client.hdel(key, fields)`](/javascript-api/k6-experimental-redis/client/client-hdel)                 | **[HDEL](https://redis.io/commands/hdel)**       | Deletes the specified fields from the hash stored at `key`.                                          |
| [`Client.hgetall(key)`](/javascript-api/k6-experimental-redis/client/client-hgetall)                   | **[HGETALL](https://redis.io/commands/hgetall)** | Returns all fields and values of the hash stored at `key`.                                           |
| [`Client.hkeys(key)`](/javascript-api/k6-experimental-redis/client/client-hkeys)                       | **[HKEYS](https://redis.io/commands/hkeys)**     | Returns all fields of the hash stored at `key`.                                                      |
| [`Client.hvals(key)`](/javascript-api/k6-experimental-redis/client/client-hvals)                       | **[HVALS](https://redis.io/commands/hvals)**     | Returns all values of the hash stored at `key`.                                                      |
| [`Client.hlen(key)`](/javascript-api/k6-experimental-redis/client/client-hlen)                         | **[HLEN](https://redis.io/commands/hlen)**       | Returns the number of fields in the hash stored at `key`.                                            |
| [`Client.hincrby(key, field, increment)`](/javascript-api/k6-experimental-redis/client/client-hincrby) | **[HINCRBY](https://redis.io/commands/hincrby)** | Increments the integer value of `field` in the hash stored at `key` by `increment`.                  |

## Set methods

| Method                                                                            | Redis command                                              | Description                                                              |
| :-------------------------------------------------------------------------------- | :--------------------------------------------------------- | :----------------------------------------------------------------------- |
| [`Client.sadd(key, members)`](/javascript-api/k6-experimental-redis/client/client-sadd)          | **[SADD](https://redis.io/commands/sadd)**               | Adds the specified members to the set stored at `key`.                   |
| [`Client.srem(key, members)`](/javascript-api/k6-experimental-redis/client/client-srem)          | **[SREM](https://redis.io/commands/srem)**               | Removes the specified members from the set stored at `key`.              |
| [`Client.sismember(key, member)`](/javascript-api/k6-experimental-redis/client/client-sismember) | **[SISMEMBER](https://redis.io/commands/sismember)**     | Returns if member is a member of the set stored at `key`.                |
| [`Client.smembers(key)`](/javascript-api/k6-experimental-redis/client/client-smembers)           | **[SMEMBERS](https://redis.io/commands/smembers)**       | Returns all the members of the set values stored at `keys`.              |
| [`Client.srandmember(key)`](/javascript-api/k6-experimental-redis/client/client-srandmember)     | **[SRANDMEMBER](https://redis.io/commands/srandmember)** | Returns a random element from the set value stored at `key`.             |
| [`Client.spop(key)`](/javascript-api/k6-experimental-redis/client/client-spop)                   | **[SPOP](https://redis.io/commands/spop)**               | Removes and returns a random element from the set value stored at `key`. |

## miscellaneous

| Method                                                                                  | Description                         |
| :-------------------------------------------------------------------------------------- | :---------------------------------- |
| [`Client.sendCommand(command, args)`](/javascript-api/k6-experimental-redis/client/client-sendCommand) | Send a command to the Redis server. || Increments the number stored at `key` by `increment`.                       |
| [`Client.decr(key)`](/javascript-api/k6-experimental-redis/client/client-decr)                  | **[DECR](https://redis.io/commands/decr)**           | Decrements the number stored at `key` by one.                               |
| [`Client.decrBy(key, decrement)`](/javascript-api/k6-experimental-redis/client/client-decrby)   | **[DECRBY](https://redis.io/commands/decrby)**       | Decrements the number stored at `key` by `decrement`.                       |
| [`Client.randomKey()`](/javascript-api/k6-experimental-redis/client/client-randomkey)           | **[RANDOMKEY](https://redis.io/commands/randomkey)** | Returns a random key's value.                                               |
|
```

**Review by MattDodsonEnglish — CHANGES_REQUESTED** [review]
In my last-second rereview, I discovered some broken links on the Client page. It looks the anchors still went to what I guess was an old path, `k6-redis`.

I think the suggestion should fix it.

**Pull request merged**
